File: lib/src/dnn/dnn_async.dart
// ignore_for_file: non_constant_identifier_names

library cv;

import 'dart:ffi' as ffi;
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/mat_type.dart';
import '../core/rect.dart';
import '../core/scalar.dart';
import '../core/size.dart';
import '../core/vec.dart';
import '../opencv.g.dart' as cvg;
import './dnn.dart';

extension LayerAsync on Layer {
  Future<int> inputNameToIndexAsync(String name) async {
    final cName = name.toNativeUtf8().cast<ffi.Char>();
    final rval = await cvRunAsync<int>(
      (callback) => CFFI.Layer_InputNameToIndex_Async(ref, cName, callback),
      (c, p) {
        final rval = p.cast<ffi.Int>().value;
        calloc.free(p);
        return c.complete(rval);
      },
    );
    calloc.free(cName);

    return rval;
  }

  Future<int> outputNameToIndexAsync(String name) async {
    final cName = name.toNativeUtf8().cast<ffi.Char>();
    final rval = cvRunAsync<int>(
      (callback) => CFFI.Layer_OutputNameToIndex_Async(ref, cName, callback),
      (c, p) {
        final rval = p.cast<ffi.Int>().value;
        calloc.free(p);
        return c.complete(rval);
      },
    );
    calloc.free(cName);
    return rval;
  }
}

extension NetAsync on Net {
  static Future<Net> emptyAsync() async {
    final rval = await cvRunAsync<Net>(CFFI.Net_Create_Async, (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });

    return rval;
  }

  static Future<Net> fromFileAsync(String path, {String config = "", String framework = ""}) async {
    final cPath = path.toNativeUtf8().cast<ffi.Char>();
    final cConfig = config.toNativeUtf8().cast<ffi.Char>();
    final cFramework = framework.toNativeUtf8().cast<ffi.Char>();
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNet_Async(cPath, cConfig, cFramework, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    calloc.free(cPath);
    calloc.free(cConfig);
    calloc.free(cFramework);

    return rval;
  }

  static Future<Net> fromBytesAsync(
    String framework,
    Uint8List bufferModel, {
    Uint8List? bufferConfig,
  }) async {
    bufferConfig ??= Uint8List(0);
    final cFramework = framework.toNativeUtf8().cast<ffi.Char>();
    final bufM = VecUChar.fromList(bufferModel);
    final bufC = VecUChar.fromList(bufferConfig);
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetBytes_Async(cFramework, bufM.ref, bufC.ref, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });

    calloc.free(cFramework);

    return rval;
  }

  static Future<Net> fromCaffeAsync(String prototxt, String caffeModel) async {
    final cProto = prototxt.toNativeUtf8().cast<ffi.Char>();
    final cCaffe = caffeModel.toNativeUtf8().cast<ffi.Char>();
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetFromCaffe_Async(cProto, cCaffe, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    calloc.free(cProto);
    calloc.free(cCaffe);

    return rval;
  }

  static Future<Net> fromCaffeBytesAsync(Uint8List bufferProto, Uint8List bufferModel) async {
    final bufP = VecUChar.fromList(bufferProto);
    final bufM = VecUChar.fromList(bufferModel);
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetFromCaffeBytes_Async(bufP.ref, bufM.ref, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    return rval;
  }

  static Future<Net> fromOnnxAsync(String path) async {
    final cpath = path.toNativeUtf8().cast<ffi.Char>();
    final rval = await cvRunAsync<Net>((callback) => CFFI.Net_ReadNetFromONNX_Async(cpath, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    calloc.free(cpath);

    return rval;
  }

  static Future<Net> fromOnnxBytesAsync(Uint8List bufferModel) async {
    final bufM = VecUChar.fromList(bufferModel);
    final rval =
        await cvRunAsync<Net>((callback) => CFFI.Net_ReadNetFromONNXBytes_Async(bufM.ref, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    return rval;
  }

  static Future<Net> fromTensorflowAsync(String path, {String config = ""}) async {
    final cpath = path.toNativeUtf8().cast<ffi.Char>();
    final cconf = config.toNativeUtf8().cast<ffi.Char>();
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetFromTensorflow_Async(cpath, cconf, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    calloc.free(cpath);
    calloc.free(cconf);

    return rval;
  }

  static Future<Net> fromTensorflowBytesAsync(Uint8List bufferModel, {Uint8List? bufferConfig}) async {
    bufferConfig ??= Uint8List(0);
    final bufM = VecUChar.fromList(bufferModel);
    final bufC = VecUChar.fromList(bufferConfig);
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetFromTensorflowBytes_Async(bufM.ref, bufC.ref, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    return rval;
  }

  static Future<Net> fromTFLiteAsync(String path) async {
    final cpath = path.toNativeUtf8().cast<ffi.Char>();
    final rval =
        await cvRunAsync<Net>((callback) => CFFI.Net_ReadNetFromTFLite_Async(cpath, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    calloc.free(cpath);
    return rval;
  }

  static Future<Net> fromTFLiteBytesAsync(Uint8List bufferModel) async {
    final bufM = VecUChar.fromList(bufferModel);
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetFromTFLiteBytes_Async(bufM.ref, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    return rval;
  }

  static Future<Net> fromTorchAsync(String path, {bool isBinary = true, bool evaluate = true}) async {
    final cpath = path.toNativeUtf8().cast<ffi.Char>();
    final rval = await cvRunAsync<Net>(
        (callback) => CFFI.Net_ReadNetFromTorch_Async(cpath, isBinary, evaluate, callback), (c, p) {
      return c.complete(Net.fromPointer(p.cast<cvg.Net>()));
    });
    calloc.free(cpath);
    return rval;
  }

  Future<bool> isEmptyAsync() async {
    final rval = cvRunAsync<bool>(
      (callback) => CFFI.Net_Empty_Async(ref, callback),
      (c, p) {
        final rval = p.cast<ffi.Bool>().value;
        calloc.free(p);
        return c.complete(rval);
      },
    );
    return rval;
  }

  Future<String> dumpAsync() async {
    final rval = cvRunAsync<String>(
      (callback) => CFFI.Net_Dump_Async(ref, callback),
      (c, p) {
        final rval = p.cast<ffi.Pointer<ffi.Char>>().value.toDartString();
        calloc.free(p);
        return c.complete(rval);
      },
    );
    return rval;
  }

  Future<void> setInputAsync(InputArray blob, {String name = ""}) async {
    final cname = name.toNativeUtf8().cast<ffi.Char>();
    await cvRunAsync0<void>(
      (callback) => CFFI.Net_SetInput_Async(ref, blob.ref, cname, callback),
      (c) {
        return c.complete();
      },
    );
    calloc.free(cname);
  }

  Future<Mat> forwardAsync({String outputName = ""}) async {
    final rval = cvRunAsync<Mat>(
      (callback) => CFFI.Net_Forward_Async(ref, outputName.toNativeUtf8().cast<ffi.Char>(), callback),
      (c, result) => c.complete(Mat.fromPointer(result.cast<cvg.Mat>())),
    );
    return rval;
  }

  Future<VecMat> forwardLayersAsync(List<String> names) async {
    final vecName = names.i8;
    final rval = cvRunAsync<VecMat>(
      (callback) => CFFI.Net_ForwardLayers_Async(ref, vecName.ref, callback),
      (c, result) => c.complete(VecMat.fromPointer(result.cast<cvg.VecMat>())),
    );
    return rval;
  }

  Future<void> setPreferableBackendAsync(int backendId) async {
    await cvRunAsync0<void>(
      (callback) => CFFI.Net_SetPreferableBackend_Async(ref, backendId, callback),
      (c) => c.complete(),
    );
  }

  Future<void> setPreferableTargetAsync(int targetId) async {
    await cvRunAsync0<void>(
      (callback) => CFFI.Net_SetPreferableTarget_Async(ref, targetId, callback),
      (c) => c.complete(),
    );
  }

  Future<int> getPerfProfileAsync() async {
    final rval = cvRunAsync<int>(
      (callback) => CFFI.Net_GetPerfProfile_Async(ref, callback),
      (c, p) {
        final rval = p.cast<ffi.Int64>().value;
        calloc.free(p);
        return c.complete(rval);
      },
    );
    return rval;
  }

  Future<VecInt> getUnconnectedOutLayersAsync() async {
    final rval = cvRunAsync<VecInt>(
      (callback) => CFFI.Net_GetUnconnectedOutLayers_Async(ref, callback),
      (c, result) => c.complete(VecInt.fromPointer(result.cast<cvg.VecInt>())),
    );
    return rval;
  }

  Future<List<String>> getLayerNamesAsync() async {
    final rval = cvRunAsync<List<String>>(
      (callback) => CFFI.Net_GetLayerNames_Async(ref, callback),
      (c, result) => c.complete(VecVecChar.fromPointer(result.cast<cvg.VecVecChar>()).asStringList()),
    );
    return rval;
  }

  Future<(VecFloat, VecInt)> getInputDetailsAsync() async {
    final rval = cvRunAsync2<(VecFloat, VecInt)>(
      (callback) => CFFI.Net_GetInputDetails_Async(ref, callback),
      (c, sc, zp) => c.complete(
        (VecFloat.fromPointer(sc.cast<cvg.VecFloat>()), VecInt.fromPointer(zp.cast<cvg.VecInt>())),
      ),
    );
    return rval;
  }

  Future<Layer> getLayerAsync(int index) async {
    final rval = cvRunAsync<Layer>(
      (callback) => CFFI.Net_GetLayer_Async(ref, index, callback),
      (c, result) => c.complete(Layer.fromPointer(result.cast<cvg.Layer>())),
    );
    return rval;
  }
}

Future<Mat> getBlobChannelAsync(Mat blob, int imgidx, int chnidx) async {
  final rval = cvRunAsync<Mat>(
    (callback) => CFFI.Net_GetBlobChannel_Async(blob.ref, imgidx, chnidx, callback),
    (c, result) => c.complete(Mat.fromPointer(result.cast<cvg.Mat>())),
  );
  return rval;
}

Future<Scalar> getBlobSizeAsync(Mat blob) async {
  final rval = cvRunAsync<Scalar>(
    (callback) => CFFI.Net_GetBlobSize_Async(blob.ref, callback),
    (c, result) => c.complete(Scalar.fromPointer(result.cast<cvg.Scalar>())),
  );
  return rval;
}

Future<List<int>> NMSBoxesAsync(
  VecRect bboxes,
  VecFloat scores,
  double scoreThreshold,
  double nmsThreshold, {
  double eta = 1.0,
  int topK = 0,
}) async {
  final rval = cvRunAsync<List<int>>(
    (callback) => CFFI.NMSBoxesWithParams_Async(
      bboxes.ref,
      scores.ref,
      scoreThreshold,
      nmsThreshold,
      eta,
      topK,
      callback,
    ),
    (c, result) => c.complete(VecInt.fromPointer(result.cast<cvg.VecInt>()).toList()),
  );
  return rval;
}

Future<Mat> blobFromImageAsync(
  InputArray image, {
  double scalefactor = 1.0,
  (int, int) size = (0, 0),
  Scalar? mean,
  bool swapRB = false,
  bool crop = false,
  int ddepth = MatType.CV_32F,
}) async {
  mean ??= Scalar.zeros;
  final rval = await cvRunAsync<Mat>(
    (callback) => CFFI.Net_BlobFromImage_Async(
      image.ref,
      scalefactor,
      size.cvd.ref,
      mean!.ref,
      swapRB,
      crop,
      ddepth,
      callback,
    ),
    (c, mat) => c.complete(Mat.fromPointer(mat.cast<cvg.Mat>())),
  );
  return rval;
}

Future<Mat> blobFromImagesAsync(
  VecMat images, {
  double scalefactor = 1.0,
  (int, int) size = (0, 0),
  Scalar? mean,
  bool swapRB = false,
  bool crop = false,
  int ddepth = MatType.CV_32F,
}) async {
  mean ??= Scalar.zeros;
  final rval = await cvRunAsync<Mat>(
    (callback) => CFFI.Net_BlobFromImages_Async(
      images.ref,
      scalefactor,
      size.cvd.ref,
      mean!.ref,
      swapRB,
      crop,
      ddepth,
      callback,
    ),
    (c, blob) => c.complete(Mat.fromPointer(blob.cast<cvg.Mat>())),
  );
  return rval;
}

Future<List<Mat>> imagesFromBlobAsync(Mat blob) async {
  final rval = cvRunAsync<List<Mat>>(
    (callback) => CFFI.Net_ImagesFromBlob_Async(blob.ref, callback),
    (c, result) => c.complete(VecMat.fromPointer(result.cast<cvg.VecMat>()).toList()),
  );
  return rval;
}

File: src/dnn/dnn_async.cpp
#include "dnn_async.h"
#include "core/types.h"
#include "core/vec.hpp"

// Asynchronous functions for Net
CvStatus *Net_Create_Async(CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net()});
  END_WRAP
}

CvStatus *Net_FromNet_Async(Net net, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(*net.ptr)});
  END_WRAP
}

CvStatus *Net_ReadNet_Async(
    const char *model, const char *config, const char *framework, CvCallback_1 callback
) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(cv_dnn::readNet(model, config, framework))});
  END_WRAP
}

CvStatus *Net_ReadNetBytes_Async(
    const char *framework, VecUChar model, VecUChar config, CvCallback_1 callback
) {
  BEGIN_WRAP
  auto _model = vecuchar_c2cpp(model);
  auto _config = vecuchar_c2cpp(config);
  callback(new Net{new cv_dnn::Net(cv_dnn::readNet(framework, _model, _config))});
  END_WRAP
}

CvStatus *
Net_ReadNetFromCaffe_Async(const char *prototxt, const char *caffeModel, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromCaffe(prototxt, caffeModel))});
  END_WRAP
}

CvStatus *
Net_ReadNetFromCaffeBytes_Async(VecUChar prototxt, VecUChar caffeModel, CvCallback_1 callback) {
  BEGIN_WRAP
  auto _prototxt = vecuchar_c2cpp(prototxt);
  auto _Model = vecuchar_c2cpp(caffeModel);
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromCaffe(_prototxt, _Model))});
  END_WRAP
}

CvStatus *
Net_ReadNetFromTensorflow_Async(const char *model, const char *config, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromTensorflow(model, config))});
  END_WRAP
}

CvStatus *
Net_ReadNetFromTensorflowBytes_Async(VecUChar model, VecUChar config, CvCallback_1 callback) {
  BEGIN_WRAP
  auto _model = vecuchar_c2cpp(model);
  auto _config = vecuchar_c2cpp(config);
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromTensorflow(_model, _config))});
  END_WRAP
}

CvStatus *Net_ReadNetFromTFLite_Async(const char *model, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromTFLite(model))});
  END_WRAP
}

CvStatus *Net_ReadNetFromTFLiteBytes_Async(VecUChar bufferModel, CvCallback_1 callback) {
  BEGIN_WRAP
  auto _Model = vecuchar_c2cpp(bufferModel);
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromTFLite(_Model))});
  END_WRAP
}

CvStatus *
Net_ReadNetFromTorch_Async(const char *model, bool isBinary, bool evaluate, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromTorch(model, isBinary, evaluate))});
  END_WRAP
}

CvStatus *Net_ReadNetFromONNX_Async(const char *model, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromONNX(model))});
  END_WRAP
}

CvStatus *Net_ReadNetFromONNXBytes_Async(VecUChar model, CvCallback_1 callback) {
  BEGIN_WRAP
  auto _model = vecuchar_c2cpp(model);
  callback(new Net{new cv_dnn::Net(cv_dnn::readNetFromONNX(_model))});
  END_WRAP
}

void Net_Close_Async(NetPtr net, CvCallback_0 callback) {
  CVD_FREE(net);
  callback();
}

CvStatus *Net_BlobFromImage_Async(
    Mat image,
    double scalefactor,
    Size size,
    Scalar mean,
    bool swapRB,
    bool crop,
    int ddepth,
    CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Size sz(size.width, size.height);
  cv::Scalar cm(mean.val1, mean.val2, mean.val3, mean.val4);
  Mat *blob = new Mat{new cv::Mat()};
  cv_dnn::blobFromImage(*image.ptr, *blob->ptr, scalefactor, sz, cm, swapRB, crop, ddepth);
  callback(blob);
  END_WRAP
}

CvStatus *Net_BlobFromImages_Async(
    VecMat images,
    double scalefactor,
    Size size,
    Scalar mean,
    bool swapRB,
    bool crop,
    int ddepth,
    CvCallback_1 callback
) {
  BEGIN_WRAP
  cv::Size sz(size.width, size.height);
  cv::Scalar cm = cv::Scalar(mean.val1, mean.val2, mean.val3, mean.val4);
  cv::Mat blob;
  auto _images = vecmat_c2cpp(images);
  cv_dnn::blobFromImages(_images, blob, scalefactor, sz, cm, swapRB, crop, ddepth);
  callback(new Mat{new cv::Mat(blob)});
  END_WRAP
}

CvStatus *Net_ImagesFromBlob_Async(Mat blob, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::Mat> imgs;
  cv_dnn::imagesFromBlob(*blob.ptr, imgs);
  callback(vecmat_cpp2c_p(imgs));
  END_WRAP
}

CvStatus *Net_Empty_Async(Net net, CvCallback_1 callback) {
  BEGIN_WRAP
  bool rval = net.ptr->empty();
  callback(new bool(rval));
  END_WRAP
}

CvStatus *Net_Dump_Async(Net net, CvCallback_1 callback) {
  BEGIN_WRAP
  auto ss = net.ptr->dump();
  callback(new char *(strdup(ss.c_str())));
  END_WRAP
}

CvStatus *Net_SetInput_Async(Net net, Mat blob, const char *name, CvCallback_0 callback) {
  BEGIN_WRAP
  net.ptr->setInput(*blob.ptr, name);
  callback();
  END_WRAP
}

CvStatus *Net_Forward_Async(Net net, const char *outputName, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Mat{new cv::Mat(net.ptr->forward(outputName))});
  END_WRAP
}

CvStatus *Net_ForwardLayers_Async(Net net, VecVecChar outBlobNames, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::Mat> blobs;
  std::vector<cv::String> names = vecvecchar_c2cpp_s(outBlobNames);
  net.ptr->forward(blobs, names);
  callback(vecmat_cpp2c_p(blobs));
  END_WRAP
}

CvStatus *Net_SetPreferableBackend_Async(Net net, int backend, CvCallback_0 callback) {
  BEGIN_WRAP
  net.ptr->setPreferableBackend(backend);
  callback();
  END_WRAP
}

CvStatus *Net_SetPreferableTarget_Async(Net net, int target, CvCallback_0 callback) {
  BEGIN_WRAP
  net.ptr->setPreferableTarget(target);
  callback();
  END_WRAP
}

CvStatus *Net_GetPerfProfile_Async(Net net, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<double> layersTimes;
  int64_t rval = net.ptr->getPerfProfile(layersTimes);
  callback(new int64_t(rval));
  END_WRAP
}

CvStatus *Net_GetUnconnectedOutLayers_Async(Net net, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<int> layers = net.ptr->getUnconnectedOutLayers();
  callback(vecint_cpp2c_p(layers));
  END_WRAP
}

CvStatus *Net_GetLayerNames_Async(Net net, CvCallback_1 callback) {
  BEGIN_WRAP
  std::vector<cv::String> cstrs = net.ptr->getLayerNames();
  callback(vecvecchar_cpp2c_s_p(cstrs));
  END_WRAP
}

CvStatus *Net_GetInputDetails_Async(Net net, CvCallback_2 callback) {
  BEGIN_WRAP
  std::vector<float> sc;
  std::vector<int> zp;
  net.ptr->getInputDetails(sc, zp);
  callback(vecfloat_cpp2c_p(sc), vecint_cpp2c_p(zp));
  END_WRAP
}

CvStatus *Net_GetBlobChannel_Async(Mat blob, int imgidx, int chnidx, CvCallback_1 callback) {
  BEGIN_WRAP
  size_t w = blob.ptr->size[3];
  size_t h = blob.ptr->size[2];
  callback(new Mat{new cv::Mat(h, w, CV_32F, blob.ptr->ptr<float>(imgidx, chnidx))});
  END_WRAP
}

CvStatus *Net_GetBlobSize_Async(Mat blob, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Scalar{
      (double)(blob.ptr->size[0]),
      (double)(blob.ptr->size[1]),
      (double)(blob.ptr->size[2]),
      (double)(blob.ptr->size[3])
  });
  END_WRAP
}

CvStatus *Net_GetLayer_Async(Net net, int layerid, CvCallback_1 callback) {
  BEGIN_WRAP
  callback(new Layer{new cv::Ptr<cv_dnn::Layer>(net.ptr->getLayer(layerid))});
  END_WRAP
}

CvStatus *Layer_InputNameToIndex_Async(Layer layer, const char *name, CvCallback_1 callback) {
  BEGIN_WRAP
  int rval = (*layer.ptr)->inputNameToIndex(name);
  callback(new int(rval));
  END_WRAP
}

CvStatus *Layer_OutputNameToIndex_Async(Layer layer, const char *name, CvCallback_1 callback) {
  BEGIN_WRAP
  int rval = (*layer.ptr)->outputNameToIndex(name);
  callback(new int(rval));
  END_WRAP
}

CvStatus *Layer_GetName_Async(Layer layer, CvCallback_1 callback) {
  BEGIN_WRAP
  auto ss = (*layer.ptr)->name;
  callback(new char *(strdup(ss.c_str())));
  END_WRAP
}

CvStatus *Layer_GetType_Async(Layer layer, CvCallback_1 callback) {
  BEGIN_WRAP
  auto ss = (*layer.ptr)->type;
  callback(new char *(strdup(ss.c_str())));
  END_WRAP
}

void Layer_Close_Async(LayerPtr layer, CvCallback_0 callback) {
  layer->ptr->reset();
  CVD_FREE(layer);
  callback();
}

CvStatus *NMSBoxes_Async(
    VecRect bboxes, VecF32 scores, float score_threshold, float nms_threshold, CvCallback_1 callback
) {
  BEGIN_WRAP
  std::vector<int> v;
  auto _bboxes = vecrect_c2cpp(bboxes);
  auto _scores = vecfloat_c2cpp(scores);
  cv_dnn::NMSBoxes(_bboxes, _scores, score_threshold, nms_threshold, v, 1.f, 0);
  callback(vecint_cpp2c_p(v));
  END_WRAP
}

CvStatus *NMSBoxesWithParams_Async(
    VecRect bboxes,
    VecF32 scores,
    const float score_threshold,
    const float nms_threshold,
    const float eta,
    const int top_k,
    CvCallback_1 callback
) {
  BEGIN_WRAP
  std::vector<int> v;
  auto _bboxes = vecrect_c2cpp(bboxes);
  auto _scores = vecfloat_c2cpp(scores);
  cv_dnn::NMSBoxes(_bboxes, _scores, score_threshold, nms_threshold, v, eta, top_k);
  callback(vecint_cpp2c_p(v));
  END_WRAP
}
